Update providing Application / Application context to Singleton dependencies, upgrade to gradle build plugin 2.3.3, upgrade to buildToolsVersion 26.0.1, and targetSdkVersion 26. Updates #42

chronvas opened the pull request:
App extends DaggerApplication, along with some minor version updates, a new lib and removal of unnecessary files.

**UPDATE (FROM VANDOLF):**

This PR provides a different, better solution to #42. This replaces the previous solution that uses `@BindsIntance`: #44

Some other things to note;

* "App extends DaggerApplication" - this PR does not make this change: https://github.com/vestrel00/android-dagger-butterknife-mvp/pull/45#issuecomment-321622962
* "removal of unnecessary files" - no files (package-info.java) removed: https://github.com/vestrel00/android-dagger-butterknife-mvp/pull/45#discussion_r131542837

vestrel00 commented on app/build.gradle:
```diff
@@ -17,16 +17,16 @@
 android {
-    compileSdkVersion 25
-    buildToolsVersion '25.0.2'
+    compileSdkVersion 26
```
Okay

vestrel00 commented on app/build.gradle:
```diff
@@ -17,16 +17,16 @@
 
         minSdkVersion 21
-        targetSdkVersion 25
+        targetSdkVersion 26
```
Cool

vestrel00 commented:
@chronvas thanks for the contribution! I would love to merge PRs from you. First, address my feedback before I merge. Thanks!

vestrel00 commented on app/src/main/java/com/vestrel00/daggerbutterknifemvp/AppComponent.java:
```diff
@@ -20,28 +20,25 @@
+        AndroidSupportInjectionModule.class,
+        AppModule.class
+})
+interface AppComponent extends AndroidInjector<App> {
```
I like this solution better than [the one I found for providing the `Application` instance](https://github.com/vestrel00/android-dagger-butterknife-mvp/pull/44#issuecomment-320503341)! This solution is more inline with Dagger 2.11. It makes use of  `AndroidInjector` and `AndroidInjector.Builder` instead of manually declaring the `inject(App app)` and `@BindsInstance` methods.

```java
@Singleton
@Component(modules = AppModule.class)
interface AppComponent extends AndroidInjector<App> {

    @Component.Builder
    abstract class Builder extends AndroidInjector.Builder<App> {
    }
}

@Module(includes = AndroidInjectionModule.class)
abstract class AppModule {

    @Binds
    @Singleton
    // Singleton annotation isn't necessary (in this case since Application instance is unique)
    // but is here for convention.
    abstract Application application(App app);
   ...
}

public class App extends Application implements HasActivityInjector {

    @Override
    public void onCreate() {
        super.onCreate();
        DaggerAppComponent.builder().create(this).inject(this);
    }
    ...
}
```

Thanks for this @chronvas! I'd definitely like to have this part merged in.

vestrel00 commented on build.gradle:
```diff
@@ -21,7 +21,7 @@ buildscript {
 
     dependencies {
-        classpath "com.android.tools.build:gradle:2.3.1"
+        classpath "com.android.tools.build:gradle:2.3.3"
```
Nice.

chronvas commented:
Hi @vestrel00 , I have addressed your feedback, please have a look.
I guess the only thing that remains is refactoring the comments above the App class, If that follows your describing blog post, or revert the class to it's previous state. 
In case of reverting let me know, and in case of refactoring the comments, I would suggest that you do that so it follows better the rest of the comments as well.

Many thanks 👍

Rainer-Lang commented:
How about using multidex? Still possible?

vestrel00 commented:
@Rainer-Lang installing multidex is still possible without having to extend `MultidexApplication`. https://stackoverflow.com/questions/26609734/how-to-enable-multidexing-with-the-new-android-multidex-support-library

I was just using `MultidexApplication` as an example (it was not a good example).

vestrel00 commented:
@chronvas I left the comment about the `App` class for the purposes of the blog / guide. I understand that most, if not all, users of *dagger.android* will just opt to extend `DaggerApplication` especially if they need to inject types other than `Activity` and `Fragment`.

However, for the purposes of the blog / guide, let's not extend `DaggerApplication` here. 

![screen shot 2017-08-10 at 1 41 25 pm](https://user-images.githubusercontent.com/4332753/29183697-a3843862-7dd1-11e7-8b60-65da90416cf8.png)

Please refactor the `App` class to the following:

```java
public class App extends Application implements HasActivityInjector {

    @Override
    public void onCreate() {
        super.onCreate();
        DaggerAppComponent.builder().create(this).inject(this);
    }
    ...
}
```

Other than that, everything looks good! I'll merge this as soon as you make the mentioned refactor.

Thanks! 👍

vestrel00 commented:
@chronvas this is great. I'm merging this now. Thanks a lot! 💯

vestrel00 commented on app/build.gradle:
```diff
@@ -44,5 +44,6 @@ dependencies {
 
     compile "com.google.dagger:dagger:$daggerVersion"
     compile "com.google.dagger:dagger-android:$daggerVersion"
+    compile "com.google.dagger:dagger-android-support:$daggerVersion"
```
I created setup that uses *dagger.android.support* with AppCompatActivity and support Fragments. Take a look at #49.

vestrel00 commented on app/src/main/java/com/vestrel00/daggerbutterknifemvp/AppComponent.java:
```diff
@@ -20,28 +20,25 @@
-    void inject(App app);
-
+@Component(modules = {
+        AndroidSupportInjectionModule.class,
```
I created setup that uses *dagger.android.support* with AppCompatActivity and support Fragments. Take a look at #49.

vestrel00 commented on app/src/main/java/com/vestrel00/daggerbutterknifemvp/App.java:
```diff
@@ -34,26 +30,15 @@
-
-    @Inject
-    DispatchingAndroidInjector<Activity> activityInjector;
+public class App extends DaggerApplication {
```
I created setup that uses *dagger.android.support* with AppCompatActivity and support Fragments. Take a look at #49.

vestrel00 commented:
FYI I created setup that uses *dagger.android.support* with AppCompatActivity and support Fragments. Take a look at #49.
